### ISSUE TITLE
No able to set `GLIBCXX_USE_CXX11_ABI` in the cpp test

### ISSUE THREAD

**Issue by JackCaoG** [pytorch/xla#3386]
## 🐛 Bug

I was working on the release wheel building and I found that cpp test build will fail at linking due to the `CXX_ABI` issue. During the release we set `CXX_ABI` to 0 and that was done by a check in [here](https://github.com/pytorch/xla/blob/master/test/cpp/CMakeLists.txt#L19). To be even more sure, I manually set [this line](https://github.com/pytorch/xla/blob/master/test/cpp/CMakeLists.txt#L66) and [this line](https://github.com/pytorch/xla/blob/master/test/cpp/CMakeLists.txt#L44) to 0 but I found something really weird.

under `test/cpp/build/CMakeFiles/test_ptxla.dir/flags.make` I saw

```
CXX_FLAGS =-g   -D_GLIBCXX_USE_CXX11_ABI=0 -Wno-sign-compare -Wno-deprecated-declarations -Wno-return-type -Wno-macro-redefined -Wno-return-std-move -fsized-deallocation -D_GLIBCXX_USE_CXX11_ABI=1
```

I commented out the
```
target_compile_options(test_ptxla PRIVATE ${TGT_OPTS})
```

rerun the build and then saw

```
CXX_FLAGS =-g   -D_GLIBCXX_USE_CXX11_ABI=1
```

This pretty much means I have no control over `D_GLIBCXX_USE_CXX11_ABI` because at somehow there will always be an  `-D_GLIBCXX_USE_CXX11_ABI=1` in the end. I also tried to move 

```
target_compile_options(test_ptxla PRIVATE ${TGT_OPTS})
```

to the end of the CMake file but those options we config still appears before `D_GLIBCXX_USE_CXX11_ABI=1`. 

@davidel Do you have any idea where are these two config coming fron?

**Comment by JackCaoG** [comment]
by manually removing the `-D_GLIBCXX_USE_CXX11_ABI=1` in `test/cpp/build/CMakeFiles/test_ptxla.dir/flags.make` I was able to build cpp test with `CXX_ABI=0`

The workaround I am thinking right now is
1. don't build CPP test for the release wheel
2. run the cmake, have a script to update `test/cpp/build/CMakeFiles/test_ptxla.dir/flags.make` then redo the make

Both solutions are not ideal through..

**Comment by davidel** [comment]
Hi @JackCaoG , you have to match that flag to everything that links to PT, otherwise you are going to have linking issues.
Why do you want to manually override?
Today we explicitly get that flag from the PT library binary, and match it to all the stuff we build.

**Comment by JackCaoG** [comment]
right, so today pytorch sets `_GLIBCXX_USE_CXX11_ABI=0` on their release build. The problem I run into is when I follow what we had today(using `import torch; print(int(torch._C._GLIBCXX_USE_CXX11_ABI))` for our tf build and test build), for some reason our c++ test still build with  `_GLIBCXX_USE_CXX11_ABI=1`.


I did some debugging above and found that there is a mystery `_GLIBCXX_USE_CXX11_ABI=1` show up in `test/cpp/build/CMakeFiles/test_ptxla.dir/flags.make`(generated by CMake). I verified that we correctly set `_GLIBCXX_USE_CXX11_ABI=0` in test CMake file but I can't find where the pending `_GLIBCXX_USE_CXX11_ABI=1` at the end of `CXX_FLAGS` was coming from.

I commented out [this line](https://github.com/pytorch/xla/blob/master/test/cpp/CMakeLists.txt#L81), rebuild the test and still see
```
CXX_FLAGS = -g   -D_GLIBCXX_USE_CXX11_ABI=1
```

under `test/cpp/build/CMakeFiles/test_ptxla.dir/flags.make` so this flag is not something we set in test CMake file.

**Comment by JackCaoG** [comment]
I verified that my local environment that `CXXFLAGS` is not being set, and I tested that this flag is not added by the CMake file. Given that it is being appended at the end of the `CXX_FLAGS` I am guessing something must happen after our call of `target_compile_options` in CMakeLists.txt but I was not able to figure out what that is..

**Comment by JackCaoG** [comment]
Ehh.. OK I found the issue. it is coming from https://github.com/pytorch/xla/pull/3282/files#diff-ee67c9755b4b36019b46fbe0f36c2a1587c9d66915ba13d78dfc46226477574cR11 @yeounoh We should not hardcode the CXX_ABI... I should of review it more carefully....

**Comment by JackCaoG** [comment]
Some change between 1/20 and now breaks the ABI=0 use case.. might easier for me to just bisect the prs...

**Comment by JackCaoG** [comment]
Well it seem like this breakage is introduced by a pytorch pr.

If I ping PyTorch/XLA to 
```
commit e0eb2a7bbc1e7dcefa4d83c88548a9dad7c3a387 (HEAD)
Author: Wonjoo Lee <wonjoo@google.com>
Date:   Fri Feb 4 10:54:32 2022 -0800

    Pull latest changes for TensorFlow (#3312)
``` 

and pytorch to 
```
commit 364055b2771ecf9b54f1d67a8bf44bb5496476d4 (HEAD, origin/gh/george-qi/21/base)
Author: Facebook Community Bot <facebook-github-bot@users.noreply.github.com>
Date:   Thu Feb 10 00:32:02 2022 -0800
```

CXX_ABI=0 PyTorch/XLA test build will fail with CMake flag. 
```
CXX_FLAGS =-D_GLIBCXX_USE_CXX11_ABI=0 -g   -D_GLIBCXX_USE_CXX11_ABI=0 -Wno-sign-compare -Wno-deprecated-declarations -Wno-return-type -Wno-macro-redefined -Wno-return-std-move -fsized-deallocation -D_GLIBCXX_USE_CXX11_ABI=1
```

If I instead ping PyTorch to a earlier date
```
commit ff8fb717db381de5bbfadd629314ababbf75c12d (HEAD)
Author: Nikita Shulga <nshulga@fb.com>
Date:   Tue Jan 18 14:07:20 2022 -0800

    Fix `get_git_repo_dir` (#71448)
```

build passed and in the CMake Flag I saw
```
jackcao@jackcao64:~/Desktop/gitProj2/pytorch/xla$ cat test/cpp/build/CMakeFiles/test_ptxla.dir/flags.make 
# CMAKE generated file: DO NOT EDIT!
# Generated by "Unix Makefiles" Generator, CMake Version 3.14

# compile CXX with /usr/bin/clang++
CXX_FLAGS = -D_GLIBCXX_USE_CXX11_ABI=0 -g   -D_GLIBCXX_USE_CXX11_ABI=0 -Wno-sign-compare -Wno-deprecated-declarations -Wno-return-type -Wno-macro-redefined -Wno-return-std-move -fsized-deallocation
```


Bisecting PyTorch prs..

**Comment by JackCaoG** [comment]
https://github.com/pytorch/pytorch/commit/1cc824ef598c7d249ff48643f9148c40988e6fdd seems to be the offending pr.

**Comment by davidel** [comment]
AFAICT, even though the following line explicitly sets the flag in the compile options (now for all **Linux** instead of **GCC**), if the PT build machinery would be consistent, we would find a matching flag with our Python script poking, and there would be no problem:

https://github.com/pytorch/pytorch/commit/1cc824ef598c7d249ff48643f9148c40988e6fdd#diff-c181594976c3ae3ca5a3c731ef2da63b409050fa968ded47203ae3cbea9ec23eR170

IMHO there is an inconsistency between the build machinery above, the the Python advertised flag.

**Comment by JackCaoG** [comment]
Right, I think what happened is that during our wheel build, in the base Dockerfile we set the env
```
ENV CXX_ABI "${cxx_abi}"
```
and then read it in our build script we do
```
  if [ "$CXX_ABI" == "0" ]; then
    export CFLAGS="${CFLAGS} -D_GLIBCXX_USE_CXX11_ABI=0"
    export CXXFLAGS="${CXXFLAGS} -D_GLIBCXX_USE_CXX11_ABI=0"
  fi
}
```

By manually setting the `CFLAGS` and `CXXFLAGS`, we control how pytorch and pytorch/xla ABI build seeting. However when we build the test, it reads upstream pytorch CMake, which is checking the `_GLIBCXX_USE_CXX11_ABI` we did not set. Therefore even if 
```
import torch; print(int(torch._C._GLIBCXX_USE_CXX11_ABI))
```
return 0, `_GLIBCXX_USE_CXX11_ABI` still default to 1 which insert a `-D_GLIBCXX_USE_CXX11_ABI=1` at the end of CMake flags.

Going forward I think maybe checking `_GLIBCXX_USE_CXX11_ABI` in our CMake file instead of running python call above is the right thing to do. I want to fix it for now to unblock the 1.11 release.

**Comment by JackCaoG** [comment]
Internal wheel build passed with above changes. Close this issue for now. Opening a new one to test remove 
```
  if [ "$CXX_ABI" == "0" ]; then
    export CFLAGS="${CFLAGS} -D_GLIBCXX_USE_CXX11_ABI=0"
    export CXXFLAGS="${CXXFLAGS} -D_GLIBCXX_USE_CXX11_ABI=0"
  fi
```

since `_GLIBCXX_USE_CXX11_ABI` should be enough to control pytorch CXXABI setting. PyTorch/XLA should just read pytorch setting by `import torch; print(int(torch._C._GLIBCXX_USE_CXX11_ABI))` in `setup.py`
